lib/linked_list.rb

```diff
@@ -16,102 +16,308 @@ def initialize
     def has_cycle
-      raise NotImplementedError
+      return false if @head.nil? || @head.next.nil?
+
+      slow = @head
+      fast = @head
+
+      until fast.nil?
+        slow = slow.next
+        fast = fast.next
+        fast = fast.next unless fast.nil?
+
+        return true if fast == slow
+      end
+
+      return false
     end
 
 
     # Additional Exercises 
+    # Time Complexity: O(1) because the algorithm takes the same amount of time with any linked list size
+    # Space Complexity: O(1) because the amount of values stored does not increase with linked list size
     # returns the value in the first node
     # returns nil if the list is empty
     def get_first
-      raise NotImplementedError
+      return nil if @head.nil?
+
+      return @head.data
     end
 
+    # Time Complexity: O(n) the method must visit each node of the list, so the method is dependant on list size
+    # Space Complexity: O(1) because the amount of values stored does not increase with linked list size
     # method that inserts a given value as a new last node in the linked list
     def add_last(value)
-      raise NotImplementedError
+      if @head.nil?
+        @head = Node.new(value, nil)
+        return
+      end
+
+      current = @head
+
+      (self.length - 1).times do 
+        current = current.next
+      end
+
+      new_node = Node.new(value, nil)
+      current.next = new_node
+      return
     end
 
+    # Time Complexity: O(n) the method must visit each node of the list, so the method is dependant on list size
+    # Space Complexity: O(1) because the amount of values stored does not increase with linked list size
     # method that returns the value of the last node in the linked list
     # returns nil if the linked list is empty
     def get_last
-      raise NotImplementedError
+      return nil if @head.nil?
+
+      current = @head
+
+      until current.next.nil?
+        current = current.next
+      end
+
+      return current.data
     end
 
+    # Time Complexity: O(n) In the worst case the method must visit each node of the list, so the method is dependant on list size
+    # Space Complexity: O(1) because the amount of values stored does not increase with linked list size
     # method to insert a new node with specific data value, assuming the linked
     # list is sorted in ascending order
     def insert_ascending(value)
-      raise NotImplementedError
+      return nil if @head.nil? || @head.next.nil?
+
+      current = @head
+
+      if value < current.data
+        self.add_first(value)
+        return
+      end
+
+      until current.next.nil?
+        current_value = current.data
+        next_value = current.next.data
+
+        if (value > current_value && value < next_value)
+          new_node = Node.new(value, current.next)
+
+          current.next = new_node
+          return
+        end
+
+        current = current.next
+      end
+
+      self.add_last(value)

```

Comment:
This makes a 2nd traversal through the loop, consider:

```suggestion
      current.next = Node.new(value)
```